### PR TITLE
Feat/solana suite

### LINKED ISSUE #10197: Solana support
I see we're missing an issue for this, so I'm setting it up so we don't miss it in the process and QA, but I don't want to do a description right now...

### PR DESCRIPTION
**Pull request by matejkriz**
Recreated https://github.com/trezor/trezor-suite/pull/10026 so CI can run

## Description

Final Solana PR to Suite. Connect already merged.

Resolve https://github.com/trezor/trezor-suite/issues/10197

### CODE REVIEW

**Comment by matejkriz** [comment]
@marekrjpolak @mroz22  pretty please review `packages/suite/src/utils/wallet/solanaUtils.ts` and `packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts` 🙏

**Comment by gabrielKerekes** [comment]
We'll take a look at the CI errors.

**Comment by gabrielKerekes** [comment]
CI errors should be fixed in https://github.com/vacuumlabs/trezor-suite/commit/c283e44b80b6f51dffa048871ce537e49990a3f4 in our fork since we don't have access here.

**Review by marekrjpolak — COMMENTED** [review]
- One tx per page, probably wrong pagination data from blockchain-link. But it seems to be on all coins, not just Solana? Not on `develop` though...
image
- Can't get receive address, missing `'solana'` case in some switch?
image
- Send button always disabled. Also, why is the fee negative? Also, the fee selection component probably doesn't have to be there at all.
image

**Comment by memoriesadrift** [comment]
> * One tx per page, probably wrong pagination data from blockchain-link. But it seems to be on all coins, not just Solana? Not on `develop` though...  

Can't replicate. I always see 2 txs per page, in line with a param defined in code. I wasn't the one who defined it, so missing context as to why it was added. Let me know if it should be reverted.
image

>     * Can't get receive address, missing `'solana'` case in some switch?

Fixed.

>     * Send button always disabled. Also, why is the fee negative? Also, the fee selection component probably doesn't have to be there at all.

Fixed.

File: suite-common/suite-config/src/settings.ts
Comment (memoriesadrift) [inline suite-common/suite-config/src/settings.ts]: TXs per page setting mentioned in my comment.

Comment (marekrjpolak) [reply suite-common/suite-config/src/settings.ts]: I haven't found any other source of this change than this branch, but it definitely shouldn't be there as it affects all the coins.

Comment (memoriesadrift) [reply suite-common/suite-config/src/settings.ts]: fix pushed 👍

File: packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts
Comment (marekrjpolak) [inline packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts]: Although I don't understand what does this blockhash parameter mean and why it differs every time I compose identical tx, I believe `DSOL` blockHash should be taken into account. But even if I replaced it, I wasn't able to send tx because of it.
image

Comment (gabrielKerekes) [reply packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts]: Fixed in https://github.com/vacuumlabs/trezor-suite/commit/0c0891edaf73283e288b0caca2e23d914e0a5ad3. The problem was that the mainnet blockhash was always being added to the transaction - even on a devnet account.

The blockhash parameter is there to prevent replay attacks. A transaction will only be accepted if the blockhash in the transaction isn't older than X blocks/minutes - I think it's currently around 2 minutes. [Official docs here](https://docs.solana.com/developing/programming-model/transactions#recent-blockhash).

File: suite-common/suite-config/src/settings.ts
Comment (PeterBenc) [reply suite-common/suite-config/src/settings.ts]: Probably got missed when dividing the changes for suite :thinking: We have set it to 2 temporarily when we had just free plan for solana quicknode. A fix was merged a while go which reverted it https://github.com/trezor/trezor-suite/pull/9915. @memoriesadrift pls make sure that the way you reverted it matches the original revert. 

EDIT: fixed in another PR

**Comment by PeterBenc** [comment]
I made a couple of fixes in this PR, feel free to cherry-pick the suite ones here.
https://github.com/trezor/trezor-suite/pull/10048

**Comment by matejkriz** [comment]
FIXED 

When I reconnect the device, connect is trying to fetch corrupted URL `https://https//solana.trezor.io`, not exactly sure what causes that, but maybe missing protocol in connect-common or blockchain-link config (or both)?
image

**Comment by matejkriz** [comment]
FIXED 

Is there a reason not to allow custom backend for Solana? 
image

File: packages/components/src/components/assets/CoinLogo/coins.ts
Comment (matejkriz) [inline packages/components/src/components/assets/CoinLogo/coins.ts]: Please define separate icon for Solana Devnet, consistent with testnet coins. It can be 

```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
    <circle cx="12" cy="12" r="12" fill="#E75F5F"></circle>
    <path
        d="M18.9279 15.7015L16.6168 18.2419C16.5666 18.2971 16.5058 18.3411 16.4382 18.3711C16.3707 18.4012 16.2978 18.4167 16.2242 18.4167H5.26831C5.21603 18.4167 5.16489 18.401 5.12118 18.3716C5.07746 18.3422 5.04307 18.3004 5.02224 18.2513C5.0014 18.2021 4.99502 18.1478 5.00389 18.095C5.01275 18.0422 5.03647 17.9932 5.07214 17.954L7.385 15.4137C7.43511 15.3586 7.4957 15.3147 7.56303 15.2847C7.63036 15.2546 7.703 15.239 7.77645 15.2389H18.7317C18.784 15.2389 18.8351 15.2545 18.8789 15.2839C18.9225 15.3133 18.9569 15.3551 18.9778 15.4043C18.9986 15.4534 19.005 15.5077 18.9961 15.5605C18.9872 15.6133 18.9635 15.6623 18.9279 15.7015ZM16.6168 10.5859C16.5666 10.5307 16.5058 10.4867 16.4382 10.4566C16.3707 10.4266 16.2978 10.4111 16.2242 10.4111H5.26831C5.21603 10.4111 5.16489 10.4268 5.12118 10.4561C5.07746 10.4855 5.04307 10.5274 5.02224 10.5765C5.0014 10.6256 4.99502 10.6799 5.00389 10.7327C5.01275 10.7855 5.03647 10.8345 5.07214 10.8737L7.385 13.4141C7.43511 13.4691 7.4957 13.513 7.56303 13.5431C7.63036 13.5732 7.703 13.5888 7.77645 13.5889H18.7317C18.784 13.5889 18.8351 13.5732 18.8789 13.5438C18.9225 13.5145 18.9569 13.4726 18.9778 13.4235C18.9986 13.3744 19.005 13.32 18.9961 13.2672C18.9872 13.2145 18.9635 13.1655 18.9279 13.1263L16.6168 10.5859ZM5.26831 8.76111H16.2242C16.2978 8.76114 16.3707 8.74565 16.4382 8.71558C16.5058 8.68551 16.5666 8.64151 16.6168 8.58633L18.9279 6.04594C18.9635 6.00677 18.9872 5.95777 18.9961 5.90496C19.005 5.85216 18.9986 5.79785 18.9778 5.74871C18.9569 5.69958 18.9225 5.65775 18.8789 5.62836C18.8351 5.59898 18.784 5.58333 18.7317 5.58333H7.77645C7.703 5.58346 7.63036 5.59903 7.56303 5.6291C7.4957 5.65916 7.43511 5.70307 7.385 5.75811L5.07273 8.2985C5.0371 8.33762 5.01339 8.38658 5.00451 8.43931C4.99562 8.49206 5.00195 8.54633 5.02271 8.59544C5.04348 8.64456 5.07777 8.68638 5.1214 8.71581C5.16502 8.74524 5.21608 8.76098 5.26831 8.76111Z"
        fill="#FFFFFF" />
</svg>
```
dsol

**Comment by sime** [comment]
Spotted some token inconsistencies.
* Token symbol is listed (e.g. USDT) in transaction list
* The token creation balance is listed in the transaction history, IIUC it shouldn't influence the account balance.

Screenshot 2023-11-24 at 12 24 10
Screenshot 2023-11-24 at 12 22 34

File: packages/suite-web/e2e/tests/settings/coins.test.ts
Comment (matejkriz) [inline packages/suite-web/e2e/tests/settings/coins.test.ts]: This is now failing because sol/dsol are not merged in FW so not available for e2e tests. Let's wait or comment it out till FW is merged 🙏  Then `sol` should be here in the list too.

File: packages/suite/src/support/messages.ts
Comment (matejkriz) [inline packages/suite/src/support/messages.ts]: please add attribute `dynamic: true,` for those, otherwise they will be reported as unused translations

File: packages/suite/src/components/wallet/Fees/Fees.tsx
Comment (matejkriz) [inline packages/suite/src/components/wallet/Fees/Fees.tsx]: Select bar with one option only is confusing (looks like weird button) and should be replaced by something else or at least hidden.
image

File: packages/suite/src/utils/wallet/solanaUtils.ts
Comment (matejkriz) [inline packages/suite/src/utils/wallet/solanaUtils.ts]: Why is this defined outside `encodeTokenTransferInstructionData` function?

**Comment by PeterBenc** [comment]
FYI, @matejkriz 

`https://https://` bug and `customBackends` for solana fixed in https://github.com/trezor/trezor-suite/pull/10048

File: packages/components/src/images/coins/sol.svg
Comment (matejkriz) [inline packages/components/src/images/coins/sol.svg:1]: Please replace it with this:
```
<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
    <g clip-path="url(#clip0_1319_18)">
        <circle cx="12" cy="12" r="12" fill="#000000"></circle>
        <g clip-path="url(#clip1_1319_18)">
            <path
                d="M18.9279 15.7015L16.6168 18.2419C16.5666 18.2971 16.5058 18.3411 16.4382 18.3711C16.3707 18.4012 16.2978 18.4167 16.2242 18.4167H5.26831C5.21603 18.4167 5.16489 18.401 5.12118 18.3716C5.07746 18.3422 5.04307 18.3004 5.02224 18.2513C5.0014 18.2021 4.99502 18.1478 5.00389 18.095C5.01275 18.0422 5.03647 17.9932 5.07214 17.954L7.385 15.4137C7.43511 15.3586 7.4957 15.3147 7.56303 15.2847C7.63036 15.2546 7.703 15.239 7.77645 15.2389H18.7317C18.784 15.2389 18.8351 15.2545 18.8789 15.2839C18.9225 15.3133 18.9569 15.3551 18.9778 15.4043C18.9986 15.4534 19.005 15.5077 18.9961 15.5605C18.9872 15.6133 18.9635 15.6623 18.9279 15.7015ZM16.6168 10.5859C16.5666 10.5307 16.5058 10.4867 16.4382 10.4566C16.3707 10.4266 16.2978 10.4111 16.2242 10.4111H5.26831C5.21603 10.4111 5.16489 10.4268 5.12118 10.4561C5.07746 10.4855 5.04307 10.5274 5.02224 10.5765C5.0014 10.6256 4.99502 10.6799 5.00389 10.7327C5.01275 10.7855 5.03647 10.8345 5.07214 10.8737L7.385 13.4141C7.43511 13.4691 7.4957 13.513 7.56303 13.5431C7.63036 13.5732 7.703 13.5888 7.77645 13.5889H18.7317C18.784 13.5889 18.8351 13.5732 18.8789 13.5438C18.9225 13.5145 18.9569 13.4726 18.9778 13.4235C18.9986 13.3744 19.005 13.32 18.9961 13.2672C18.9872 13.2145 18.9635 13.1655 18.9279 13.1263L16.6168 10.5859ZM5.26831 8.76111H16.2242C16.2978 8.76114 16.3707 8.74565 16.4382 8.71558C16.5058 8.68551 16.5666 8.64151 16.6168 8.58633L18.9279 6.04594C18.9635 6.00677 18.9872 5.95777 18.9961 5.90496C19.005 5.85216 18.9986 5.79785 18.9778 5.74871C18.9569 5.69958 18.9225 5.65775 18.8789 5.62836C18.8351 5.59898 18.784 5.58333 18.7317 5.58333H7.77645C7.703 5.58346 7.63036 5.59903 7.56303 5.6291C7.4957 5.65916 7.43511 5.70307 7.385 5.75811L5.07273 8.2985C5.0371 8.33762 5.01339 8.38658 5.00451 8.43931C4.99562 8.49206 5.00195 8.54633 5.02271 8.59544C5.04348 8.64456 5.07777 8.68638 5.1214 8.71581C5.16502 8.74524 5.21608 8.76098 5.26831 8.76111Z"
                fill="url(#paint0_linear_1319_18)" />
        </g>
    </g>
    <defs>
        <linearGradient id="paint0_linear_1319_18" x1="6.18176" y1="18.7225" x2="17.9883"
            y2="5.73683" gradientUnits="userSpaceOnUse">
            <stop offset="0.08" stop-color="#9945FF" />
            <stop offset="0.3" stop-color="#8752F3" />
            <stop offset="0.5" stop-color="#5497D5" />
            <stop offset="0.6" stop-color="#43B4CA" />
            <stop offset="0.72" stop-color="#28E0B9" />
            <stop offset="0.97" stop-color="#19FB9B" />
        </linearGradient>
    </defs>
</svg>
```
sol
that is more aligned with Solana brand guidelines and requested by @Hermez-cz

**Comment by sime** [comment]
NFT transaction appears as unknown. What options do we to handle that?

TX ID: 4bhkzyFxsWb193FcCuoDZyMk5TwJnJoc5WQJsufmK6PAEtk3WtCNmNpDLCGurZCZ2VaJtnDuFAQyzroQzSCvHYrW

Screenshot 2023-11-24 at 17 24 37


Screenshot 2023-11-24 at 17 22 46

**Comment by matejkriz** [comment]
FIXED by removing the button

If I click "Add token" on Solana account without tokens, Suite leads me to add ERC20 tokens.  There is no such thing on solana, right?  
image
image

**Comment by FreeWall** [comment]
Some fixes for SOL in coinmarket https://github.com/trezor/trezor-suite/pull/10072 :pray:

File: suite-native/qr-code/src/components/XpubQRCodeBottomSheet.tsx
Comment (gabrielKerekes) [inline suite-native/qr-code/src/components/XpubQRCodeBottomSheet.tsx]: Not sure what this does but it IMO should say `Receive address` or just `Public key` for Solana (they're the same thing just with different encoding).

File: packages/suite/src/utils/wallet/solanaUtils.ts
Comment (memoriesadrift) [reply packages/suite/src/utils/wallet/solanaUtils.ts]: Good catch, moving it inside the function.

**Comment by memoriesadrift** [comment]
> NFT transaction appears as unknown. What options do we to handle that?
> 
> TX ID: 4bhkzyFxsWb193FcCuoDZyMk5TwJnJoc5WQJsufmK6PAEtk3WtCNmNpDLCGurZCZ2VaJtnDuFAQyzroQzSCvHYrW

I would avoid implementing that at this stage. Displaying this TX would involve adding parsing logic for instructions of the "Bubblegum" program. Additionally, other nfts interacting with different programs (e.g. minted via Candy Machine or some different wrapper program) would not be displayed even after adding support for the Bubblegum program. Basically, every custom program's instructions would need to be checked and parsed by us.
FWIW: we already support *native* NFTs by default, as they behave and interact with the same program as SPL tokens.

**Comment by memoriesadrift** [comment]
Fixes addressed on our fixes branch: https://github.com/vacuumlabs/trezor-suite/tree/feat/solana-fixes starting with `c23b93676d2c7de1d33a8b5a9a8314acfbc251d9`

**Comment by matejkriz** [comment]
> Fixes addressed on our fixes branch: https://github.com/vacuumlabs/trezor-suite/tree/feat/solana-fixes starting with `c23b93676d2c7de1d33a8b5a9a8314acfbc251d9`

Thanks, pushed here.

**Comment by matejkriz** [comment]
> Some fixes for SOL in coinmarket #10072 🙏

merged here

**Comment by matejkriz** [comment]
Thanks for all the fixes 🖤

@memoriesadrift could you please check the failing solana unit test? `yarn run  test:unit`

**Comment by PeterBenc** [comment]
FYI, PR which optimizes the number of requests 
https://github.com/trezor/trezor-suite/pull/10138

**Comment by matejkriz** [comment]
are we worried about this warning?
image

File: packages/suite/src/views/wallet/send/index.tsx
Comment (matejkriz) [inline packages/suite/src/views/wallet/send/index.tsx:1]: nitpick?: would be nice to hide <Options> if there are none for solana
image

**Comment by memoriesadrift** [comment]
> are we worried about this warning? image

This is raised by us in connect/FW. cc: @gabrielKerekes

File: packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts
Comment (matejkriz) [inline packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts]: @memoriesadrift you can't use `boolean &&` like this, otherwise "false" is rendered, see 
image

```suggestion
                            ? `${destinationAddress} (Token account for: ${output.address}${
                                  isCreatingAccount ? ' , Account will be created.' : ''
```

File: suite-native/qr-code/src/components/XpubQRCodeBottomSheet.tsx
Comment (vytick) [reply suite-native/qr-code/src/components/XpubQRCodeBottomSheet.tsx]: `Receive address`

File: yarn.lock
Comment (vytick) [inline yarn.lock]: Please, can we have one version of Solana/web3.js instead of multiple? Lets use the same that is being used elsewhere

File: packages/suite/src/views/wallet/send/components/Raw.tsx
Comment (vytick) [inline packages/suite/src/views/wallet/send/components/Raw.tsx]: There is no rule for `0x` prefix in Solana account addresses

File: packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts
Comment (memoriesadrift) [reply packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts]: Oops, react brain got to me.

**Comment by memoriesadrift** [comment]
Addressed feedback here: https://github.com/vacuumlabs/trezor-suite/pull/29
Basing off of https://github.com/trezor/trezor-suite/pull/10164 as I expect that to be merged here beforehand.

File: yarn.lock
Comment (Nodonisko) [reply yarn.lock]: maybe `yarn dedupe` will solve this

Comment (memoriesadrift) [reply yarn.lock]: yep, it did

File: packages/connect/src/data/config.ts
Comment (matejkriz) [inline packages/connect/src/data/config.ts]: TODO: fix before merge

File: packages/suite-web/e2e/tests/settings/coins.test.ts
Comment (matejkriz) [reply packages/suite-web/e2e/tests/settings/coins.test.ts]: TODO

File: packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts
Comment (matejkriz) [inline packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts]: TODO remove

File: suite-common/wallet-config/src/networksConfig.ts
Comment (matejkriz) [inline suite-common/wallet-config/src/networksConfig.ts]: TODO

File: suite-common/wallet-config/src/networksConfig.ts
Comment (matejkriz) [inline suite-common/wallet-config/src/networksConfig.ts]: TODO

**Comment by gabrielKerekes** [comment]
Since FW not displays the base address I think we can now remove the "(Token account for: ..." message and just display the base address? Or do we maybe also want to display the token address?

cc @Hannsek @memoriesadrift

**Comment by Hannsek** [comment]
FW displays the base address. What do you mean by "(Token account for: …"? Can you provide us with screenshot please?

**Comment by gabrielKerekes** [comment]
Oops, forgot to attach screenshot sorry. In Suite in the transfer token confirmation dialog we still display the following:
Screenshot 2023-12-05 at 10 54 32

**Comment by Hannsek** [comment]
Ah, ok, then please remove it as user will confirm only the base address. Thanks!

File: packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts
Comment (matejkriz) [reply packages/suite/src/actions/wallet/send/sendFormSolanaActions.ts:80]: TODO: remove

File: packages/connect/src/data/config.ts
Comment (memoriesadrift) [reply packages/connect/src/data/config.ts]: Should we address all these todos or will you be doing it? @matejkriz

**Comment by memoriesadrift** [comment]
> Ah, ok, then please remove it as user will confirm only the base address. Thanks!

Fixed here: https://github.com/vacuumlabs/trezor-suite/pull/29/commits/42f5b67594887ba1743d3be8e11a52ba6ea2e00b

File: packages/protobuf/messages.json
Comment (matejkriz) [inline packages/protobuf/messages.json]: TODO: regenerate?

File: packages/protobuf/src/messages.ts
Comment (matejkriz) [inline packages/protobuf/src/messages.ts]: TODO: regenerate file?

Comment (memoriesadrift) [reply packages/protobuf/src/messages.ts]: Should we address all these todos or will you be doing it? @matejkriz

File: packages/connect/src/data/config.ts
Comment (matejkriz) [reply packages/connect/src/data/config.ts]: I'll do it...

**Comment by matejkriz** [comment]
@SocketSecurity ignore @expo/fingerprint@0.0.1-canary-20231205-250b31f
@SocketSecurity ignore jest-expo@50.0.0-canary-20231205-250b31f

File: packages/suite/src/components/wallet/Fees/Fees.tsx
Comment (matejkriz) [inline packages/suite/src/components/wallet/Fees/Fees.tsx]: this breaks UI for custom fee on other coins, I'll rather revert it.

Comment (matejkriz) [reply packages/suite/src/components/wallet/Fees/Fees.tsx]: af13bff35bad199567fcb4afbb162dfea23ebf01

**Comment by matejkriz** [comment]
unit test failing, @gabrielKerekes please confirm that updating the fixture would be sufficient 🙏 

failing test: https://github.com/trezor/trezor-suite/actions/runs/7102306432/job/19332347993?pr=10032#step:19:6454

fixup https://github.com/trezor/trezor-suite/pull/10032/commits/07509e9e0cd6109e8ec72df4571792478eecf32d

**Comment by matejkriz** [comment]
This PR is currently blocked by firmware PR https://github.com/trezor/trezor-firmware/pull/3359 - waiting for protobuf definitions to be in trezor-common

**Comment by gabrielKerekes** [comment]
> unit test failing, @gabrielKerekes please confirm that updating the fixture would be sufficient 🙏

You fixed it correctly, thanks - there was a mixup with that test case, I broke it yesterday, Samo based off broken code and tried to fix it and then I fixed it today but the fixture remained broken.

**Review by matejkriz — COMMENTED** [review]
Just waiting for the pipelines to pass...

**Review by tomasklim — APPROVED** [review]
Approving on behalf of @matejkriz

**Pull request merged**
